### PR TITLE
Add instrumentation API and native OpenTelemetry implementation

### PR DESCRIPTION
**Pull request by AlexanderWert**
Adds an instrumentation API with request & response lifecycle callbacks, and an implementation based on OpenTelemetry that is used by default.

To be able to implement the [OpenTelemetry semantic conventions](https://opentelemetry.io/docs/specs/semconv/database/elasticsearch/#url-path-parts), this PR also adds a `pathParameters` method to `Endpoint` that provides access to the values in URL template placeholders.

### CODE REVIEW

**Review by swallez — COMMENTED** [review]
Thanks! Overall it looks nice.

Rather than the route (i.e. path template), using the API endpoint identifier would make things easier. This identifier comes from the API specification (see e.g. [IndexRequest](https://github.com/elastic/elasticsearch-specification/blob/7ddbec9091a8aa2c7423e2df14f1c3394f421627/specification/_global/index/IndexRequest.ts#L36)) and is available as `endpoint.id()`.

Benefits:
* a single endpoint can have several routes and different http methods depending on which optional path parameters are set,
* using endpoint id avoids doing some pattern matching on the route for things like document id extraction.
* naming the span with the endpoint id ensures a 1:1 mapping from APIs to span names regardless of the path parameters set.


Note: I suggested this to @estolfo in a recent discussion.

File: java-client/src/main/java/co/elastic/clients/transport/Endpoint.java [hunk 1]
```diff
@@ -56,6 +56,18 @@
    */
   String requestUrl(RequestT request);
 
+  /**
+   * Get the route for a request (i.e. URL pattern).
+   */
+  String route(RequestT request);
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/src/main/java/co/elastic/clients/transport/Endpoint.java]: agreed on using the endpoint identifier for the span name.

File: java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class InstrumentationUtil {
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:39]: done

File: java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        Span span = tracer.spanBuilder(httpMethod + " " + route).setSpanKind(SpanKind.CLIENT).startSpan();
+        if (span.isRecording()) {
+            span.setAttribute(SemanticAttributes.DB_SYSTEM, "elasticsearch");
+            span.setAttribute(SemanticAttributes.HTTP_METHOD, endpoint.method(request));
+            span.setAttribute("url.path", endpoint.requestUrl(request));
+
+            Map<String, String> queryParameters = endpoint.queryParameters(request);
+            if (!queryParameters.isEmpty()) {
+                String queryString =
+                        queryParameters.entrySet().stream().map(e -> e.getKey() + "=" + e.getValue()).collect(Collectors.joining("&"));
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:67]: not relevant anymore

File: java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    private <RequestT> boolean shouldCaptureBody(@Nullable Span span, RequestT request, Endpoint<RequestT, ?, ?> endpoint) {
+        if (span == null || !span.isRecording()) {
+            return false;
+        }
+
+        String route = endpoint.route(request);
+
+        // We capture the request body in the span only for search-type requests.
+        return route.contains("/_search") ||
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:122]: done.

File: java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    protected void captureHostInformation(@Nullable Span span, HttpHost host) {
+        if(span == null){
+            return;
+        }
+        span.setAttribute("server.address", host.getHostName());
+        span.setAttribute("server.port", host.getPort());
+        span.setAttribute("url.scheme", host.getSchemeName());
+    }
+
+    protected <RequestT> void captureBody(@Nullable Span span, RequestT request, Endpoint<RequestT, ?, ?> endpoint,
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java]: done

File: java-client/build.gradle.kts [hunk 6]
```diff
@@ -201,6 +202,13 @@ dependencies {
     // https://github.com/eclipse-ee4j/parsson
     api("org.eclipse.parsson:parsson:1.0.0")
 
+    // OpenTelemetry API for native instrumentation of the client.
+    // Apache 2.0
+    // https://github.com/open-telemetry/opentelemetry-java
+    implementation("io.opentelemetry", "opentelemetry-api", openTelemetryVersion)
+    implementation("io.opentelemetry", "opentelemetry-semconv", "$openTelemetryVersion-alpha")
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply java-client/build.gradle.kts:209]: This is only used for OTel attribute names.

**Comment by AlexanderWert** [comment]
@swallez 

As https://github.com/open-telemetry/semantic-conventions/pull/23 has been merged, I updated this PR accordingly and also incorporated your suggestions.
I think it's safe now to finalize this PR.

**Review by swallez — APPROVED** [review]
LGTM

**Pull request merged**